### PR TITLE
 add equality for all classes except Future

### PR DESCRIPTION
**Pull request by BattiniSandeep**
Hii @vickumar1981 I have added equality for all classes except Future. I have added related test cases too. Let me know if you need any other changes.

Sandeep Kumar

### CODE REVIEW

File: pyeffects/Either.py [hunk 1]
```diff
@@ -85,7 +85,16 @@ def __init__(self, value: A) -> None:
 
     def left(self):
         return self.value
-
+    
+    def __eq__(self, __value: object) -> bool:
```
Comment (vickumar1981) [inline pyeffects/Either.py]: why the double underscore?  can we just rename this to `other`.  Also, the type should be `Either[A]` or something like that.

File: pyeffects/Either.py [hunk 2]
```diff
@@ -85,7 +85,16 @@ def __init__(self, value: A) -> None:
...
+        isLeft = isinstance(__value, self.__class__)
```
Comment (vickumar1981) [inline pyeffects/Either.py]: you can check this using the `is_right()` and `is_left()` functions.

The implementation could be something as simple as:

```
return self.is_left() == other.is_left && self.is_right() == self.is_right && self.value == other.value
```

File: pyeffects/Option.py [hunk 3]
```diff
@@ -84,6 +84,15 @@ def __init__(self, value: A) -> None:
         self.value = value
         self.biased = True
 
+    def __eq__(self, __value: object) -> bool:
```
Comment (vickumar1981) [inline pyeffects/Option.py]: ditto here.  the variable name, let's change it to `other` and change the return type from `object` to `Option[A]`.

File: tests/test_either.py [hunk 4]
```diff
@@ -62,3 +62,9 @@ def test_either_flat_map_requires_callable(self):
     def test_either_repr(self):
         assert str(Right(random_int())).startswith("Right")
         assert str(Left(random_int())).startswith("Left")
+
+    def test_left_equality(self):
```
Comment (vickumar1981) [inline tests/test_either.py:66]: also, might want to add some tests for inequality.  This is only testing what we call the "happy path".  Would probably want a test checking that equality can also be `False`.

**Comment by vickumar1981** [comment]
The build is failing with the errors below (run `pipenv run mypy pyeffects/*.py` to check the type hints)

```
pyeffects/Try.py:187: error: "object" has no attribute "value"  [attr-defined]
pyeffects/Try.py:208: error: "object" has no attribute "value"  [attr-defined]
pyeffects/Option.py:91: error: "object" has no attribute "value"  [attr-defined]
pyeffects/Option.py:1[12](https://github.com/vickumar1981/pyeffects/actions/runs/6396612398/job/17362773336?pr=32#step:6:13): error: "object" has no attribute "value"  [attr-defined]
pyeffects/Either.py:93: error: "object" has no attribute "value"  [attr-defined]
pyeffects/Either.py:1[17](https://github.com/vickumar1981/pyeffects/actions/runs/6396612398/job/17362773336?pr=32#step:6:18): error: "object" has no attribute "value"  [attr-defined]
```

This is b/c in all the functions we said that the type annotation of the input parameter was `object` instead of what class it really is, i.e. (`Future[A]`, or `Either[A]`, or `Option[A]`).

If we change that to the appropriate class, `mypy` won't complain anymore.  `object` doesn't have an attribute called `value`, but our classes do.

File: pyeffects/Either.py (hunk 1 above)
Comment (vickumar1981) [reply pyeffects/Either.py]: i.e.,

`def __eq__(self, other: 'Either[A]') -> bool:`

also, can we add some documentation to the functions as well.

https://github.com/vickumar1981/pyeffects/blob/master/pyeffects/Either.py#L37

Normally for the functions, we like to document the parameters and the return type and provide an example usage.

**Comment by vickumar1981** [comment]
@BattiniSandeep   left some comments.  thanks again for working on this!  much appreciated.

**Comment by BattiniSandeep** [comment]
Hii @vickumar1981 I have made changes acoording to our convo, But still getting error. Not sure where I'm wrong.

File: pyeffects/Try.py [hunk 5]
```diff
@@ -192,6 +195,10 @@ def __init__(self, value: A) -> None:
         self.value = value
         self.biased = True
 
+    def __eq__(self, other: 'Try[A]') -> bool:
```
Comment (vickumar1981) [inline pyeffects/Try.py]: i think you have this in here twice.

**Comment by vickumar1981** [comment]
@BattiniSandeep  yeah, i guess changing away from `object` violates the liskov subtitution of the `__eq__` method.  I think you can just add a `type: ignore` at the end of the function signature to surpress the mypy warnings.

I.e. (for try, would also have to add it to either and option):

```
def __eq__(self, other: 'Try[A]') -> bool:  # type: ignore
```

thanks!

File: pyeffects/Try.py (hunk 5 above)
Comment (BattiniSandeep) [reply pyeffects/Try.py]: I didn't get you. :)

**Comment by vickumar1981** [comment]
@BattiniSandeep   will take a look at this in a couple of days.  sorry for the delay.

**Comment by vickumar1981** [comment]
@BattiniSandeep  Pulled in all your commits in this PR and made some updates: https://github.com/vickumar1981/pyeffects/pull/34

**Comment by vickumar1981** [comment]
Closing.

**Pull request closed without merging**